### PR TITLE
feat: add Manifest.set_split_dates for absolute-date train/val/test splits

### PR DESCRIPTION
**Pull request by mikkokotila**
## What does this PR change?

Adds an alternative to the existing ratio-based `set_split_config` that pins train, val, and test windows to absolute datetime bounds. SFDs that need an exact temporal boundary (e.g. honouring a deployment date with no look-ahead leak) can opt in; existing SFDs keep working unchanged.

### Surface

```python
Manifest.set_split_dates(
    train_start, train_end,
    val_start,   val_end,
    test_start,  test_end,
) -> Manifest
```

Windows are half-open `[start, end)`. Setter validates `train_start ≤ train_end ≤ val_start ≤ val_end ≤ test_start ≤ test_end` (non-decreasing, gaps allowed); rows that fall into a gap are intentionally excluded from all three splits.

### Why date-based when `set_split_config` exists

`split_sequential` is row-count proportional and runs *after* feature engineering. Feature trim (e.g. ~115 bars on a `bbands_period=115` SFD) shifts the post-feature row count, which shifts the ratio-derived boundary, which means train+val can include rows from beyond the user's intended test boundary. That's a real leak when an SFD pins a deployment date. The new path splits by `datetime` value after features, so the boundary lands where it's specified regardless of how many rows features trimmed.

### What this PR is *not*

Not a replacement for `set_split_config`. The ratio-based path stays the ergonomic default for "I just want an 8:1:2 split on whatever data shape". The two paths are independent attributes on `Manifest` (`split_config` and `split_dates`); the new path is opt-in.

### How the safety properties carry over

The existing splitter's downstream invariants live in `MLManifest.prepare_data`'s per-split loop, not in `split_sequential` itself:

- per-split feature atomicity (`for i, split in enumerate(split_data): ...`)
- one-way scaler / target fit (`i == 0` gate)
- column alignment via `_align_split_columns`
- `_compute_alignment` reading `split_data[2]`

The new splitter produces the same `[train, val, test]` list shape, so every one of those invariants applies automatically. The blast radius of this PR is: one new manifest attribute, one new builder, one new free function, and a 3-line branch refactored into a private `_resolve_split` helper called at the two existing splitter call sites.

### Files

| File | Change | Lines |
|---|---|---|
| `limen/data/utils/splits.py` | new `split_by_dates(data, ...)` | +40 |
| `limen/data/utils/__init__.py` | export | +2 |
| `limen/experiment/manifest_core.py` | `split_dates` attribute, `set_split_dates` builder, `_resolve_split` helper, branch at the two splitter call sites, `itertools.pairwise` import | +85 |
| `tests/test_splits.py` | 10 new tests | +138 |
| `tests/run.py` | register the new tests | +20 |
| `pyproject.toml` | `3.1.1` → `3.2.0` (additive, no behaviour change to existing SFDs) | +1 |
| `CHANGELOG.md` | v3.2.0 entry | +4 |

### Tests

10 new in `tests/test_splits.py`, all registered in `tests/run.py`:

- `test_split_by_dates_returns_three_dataframes_in_order`
- `test_split_by_dates_no_row_loss_when_windows_are_contiguous`
- `test_split_by_dates_no_row_overlap_at_window_boundary`
- `test_split_by_dates_drops_rows_in_gaps_between_windows`
- `test_set_split_dates_stores_bounds_in_order`
- `test_set_split_dates_rejects_out_of_order_bounds`
- `test_set_split_dates_rejects_val_before_train_end`
- `test_set_split_dates_allows_gaps_between_adjacent_windows`
- `test_set_split_dates_does_not_mutate_split_config`
- `test_manifest_default_split_dates_is_none`

All 10 pass locally plus the 2 pre-existing `test_splits` tests; `ruff check` clean on the touched files.

## Checklist

- [x] I have reviewed full diff in "Files changed"
- [x] I left no unnecessary files in the changes
- [x] I ran `python tests/run.py` locally without errors (where applicable) — ran the touched tests directly (full suite needs the SSH tunnel)
- [ ] I updated `/docs` (this is a new public API but the existing `/docs/Experiment-Manifest.md` predates the MLManifest split; happy to add a section in a follow-up if desired)
- [x] I added and/or updated docstrings as per Writing Docstrings (for any changed public functions/classes)
- [x] I updated CHANGELOG.md
- [x] I updated pyproject.toml
- [x] I added and/or updated tests
- [x] I validated changes manually
- [x] I validated changes with LLM
- [x] I removed any extraneous examples/comments
- [ ] I linked issue to auto-close on merge — no issue; downstream consumer is Vaquum/experiment_runner (will follow with a runner PR that uses this)

### CODE REVIEW

File: limen/experiment/manifest_core.py [hunk 1]
```diff
@@ -557,7 +619,7 @@ def compute_test_bars(self, raw_data: pl.DataFrame, round_params: dict[str, Any]
             resolved = _resolve_params(base_params, round_params)
             raw_data = func(raw_data, **resolved)
 
-        split_data = split_sequential(raw_data, self.split_config)
+        split_data = _resolve_split(self, raw_data)
```
Comment (bit-mis) [inline limen/experiment/manifest_core.py]: **P2 — perf regression in `compute_test_bars` on the date-split path.**

When `split_dates` is set, `_resolve_split` delegates to `split_by_dates`, which runs three independent `data.filter(...)` passes over `raw_data` (each O(N)). `compute_test_bars` consumes only `split_data[2]`; the train and val DataFrames are built and immediately discarded.

The pre-PR `split_sequential` path was zero-copy (`data.slice` is O(1) in polars), so discarding two of three results cost nothing. The new path makes `compute_test_bars` pay ~3× the filter work for the single test slice it actually needs, and the `Log` system invokes this per round (`limen/log/log.py:94`) during result reconstruction — so the waste compounds over runs.

Simplest fix is to filter only the test window here and keep `_resolve_split` for `_run_prepare_setup` (which legitimately needs all three):

```python
if self.split_dates is not None:
    *_, test_start, test_end = self.split_dates
    test_split = raw_data.filter(
        (pl.col('datetime') >= test_start) & (pl.col('datetime') < test_end)
    )
else:
    test_split = split_sequential(raw_data, self.split_config)[2]
```

**Review by bit-mis — CHANGES_REQUESTED** [review]
Change requested in inline comments.

File: limen/experiment/manifest_core.py (hunk 1 above)
Comment (zero-bang) [reply limen/experiment/manifest_core.py]: +1 — agreed. The three-pass filter is wasted work here since only the test slice is consumed. The inline fix you propose is clean and keeps `_resolve_split` for the path that genuinely needs all three splits.

Comment (mikkokotila) [reply limen/experiment/manifest_core.py]: Fixed in 16d7798. compute_test_bars now branches inline: on the date path it filters only the test window (one O(N) pass, no discarded materialisations); on the ratio path it keeps the zero-copy slice[2]. _resolve_split stays as-is for _run_prepare_setup which legitimately needs all three slices. Two regression tests (test_compute_test_bars_returns_test_window_on_date_path and test_compute_test_bars_still_works_on_ratio_path) pin both paths.

File: limen/experiment/manifest_core.py [hunk 2]
```diff
@@ -227,6 +229,7 @@ class Manifest:
     test_data_source_config: DataSourceConfig = None
     pre_split_data_selector: PipelineStep = None
     split_config: tuple[int, int, int] = (8, 1, 2)
+    split_dates: tuple | None = None
```
Comment (zero-bang) [inline limen/experiment/manifest_core.py:233]: **`with_params_override(split_config=...)` becomes silently ineffective when `split_dates` is set.**

`_resolve_split` and the inlined `compute_test_bars` branch both check `split_dates` first — if non-None, `split_config` is never consulted. But `with_params_override` (line 575) happily accepts and sets a `split_config` override without clearing `split_dates`. So:

```python
m = manifest.set_split_dates(...)
m2 = m.with_params_override(split_config=(6, 2, 2))
# m2.split_config is (6, 2, 2) but runtime still uses split_dates
```

The override succeeds without error yet has no effect — a silent semantic trap. The trainer's round-level overrides flow through `with_params_override`, so this is a reachable path.

Suggested fix — clear `split_dates` when `split_config` is overridden:

```python
if 'split_config' in overrides:
    # ... existing validation ...
    new_manifest.split_config = sc
    new_manifest.split_dates = None   # ratio override supersedes date split
```

**Review by zero-bang — CHANGES_REQUESTED** [review]
Clean, well-tested, well-documented addition. The split function is correct, the half-open window semantics are sound, the validation is tight, and the `compute_test_bars` perf fix (addressed since @bit-mis's feedback) is the right call. One remaining issue on the `with_params_override` interaction.

File: limen/experiment/manifest_core.py [hunk 3]
```diff
@@ -1307,6 +1379,25 @@ def _run_prepare_setup(
     return split_data, all_datetimes, price_data_for_backtest
 
 
+def _resolve_split(manifest: 'Manifest', raw_data: pl.DataFrame) -> list[pl.DataFrame]:
+
+    '''
+    Pick between the date-based and ratio-based splitters.
+
+    When `manifest.split_dates` is set (via `set_split_dates`), it takes
```
[the remark this answers is not in the document]
Comment (mikkokotila) [reply limen/experiment/manifest_core.py:1407]: Fixed in a71d480. `with_params_override` now clears `split_dates` when `split_config` is in the override map — before the new `split_config` is assigned. The override comment in-place explains the silent-shadow concern (Trainer.train_sensors (1,0,0) retraining path). Pinned with `test_with_params_override_split_config_clears_split_dates` and a companion test that confirms `split_dates` is *not* cleared when only data-source params are overridden.

File: limen/experiment/manifest_core.py [hunk 4]
```diff
@@ -434,6 +437,65 @@ def set_split_config(self, train: int, val: int, test: int) -> 'Manifest':
...
+        Pin train, val, and test windows to absolute `datetime` bounds.
+
+        When set, takes precedence over `set_split_config` ratios in
+        `prepare_data` and `compute_test_bars`: the split is performed by
+        filtering the post-feature data by `datetime`, so the boundary
+        lands exactly on the datetime values regardless of how many bars
+        upstream feature engineering trims. Use when temporal boundaries
+        must be exact (e.g. honouring a deployment date with no
+        look-ahead leak); use `set_split_config` for the more ergonomic
+        ratio-based split when exact dates are not required.
```
[the remark this answers is not in the document]
Comment (mikkokotila) [reply limen/experiment/manifest_core.py]: Fixed in a71d480. The docstring on `set_split_dates` now says explicitly: 'the split runs on the raw data (in `_run_prepare_setup`, before the per-split feature transforms) and each row's slice assignment is fixed by which window its `datetime` falls into. Subsequent per-split feature engineering can trim rows independently within each slice but cannot move rows between slices, so the train/val/test datetime bounds are honoured exactly.' The `_resolve_split` and `split_by_dates` docstrings carry the same corrected mechanism.

File: limen/experiment/manifest_core.py [hunk 5]
```diff
@@ -1307,6 +1379,25 @@ def _run_prepare_setup(
     return split_data, all_datetimes, price_data_for_backtest
 
 
+def _resolve_split(manifest: 'Manifest', raw_data: pl.DataFrame) -> list[pl.DataFrame]:
+
+    '''
+    Pick between the date-based and ratio-based splitters.
+
+    When `manifest.split_dates` is set (via `set_split_dates`), it takes
+    precedence and the boundary lands exactly on the datetime values
```
[the remark this answers is not in the document]
Comment (mikkokotila) [reply limen/experiment/manifest_core.py]: Fixed in a71d480. Two new integration tests drive `MLManifest.prepare_data` end-to-end on both branches:
- `test_prepare_data_honours_split_dates_at_real_call_site` runs the full `_run_prepare_setup` → `_resolve_split` → `_finalize_to_data_dict` path with `split_dates` set, asserts the per-slice row counts match the configured windows exactly, and asserts the `_alignment` metadata (`first_test_datetime`, `last_test_datetime`, empty `missing_datetimes`) is correctly produced.
- `test_prepare_data_ratio_path_unchanged_when_split_dates_unset` pins the ratio path through the same call site so the date-path branch does not regress the default.

File: limen/experiment/manifest_core.py [hunk 6]
```diff
@@ -434,6 +437,65 @@ def set_split_config(self, train: int, val: int, test: int) -> 'Manifest':
 
         return self
 
+    def set_split_dates(
+        self,
+        train_start: object, train_end: object,
+        val_start: object, val_end: object,
+        test_start: object, test_end: object,
+    ) -> 'Manifest':
```
[the remark this answers is not in the document]
Comment (mikkokotila) [reply limen/experiment/manifest_core.py:446]: Fixed in a71d480. `set_split_dates` now annotates each bound as `date` and validates `isinstance(value, date)` at the API boundary (which catches `datetime` too, since `datetime` is a subclass). Non-date/datetime values raise `TypeError` with the bound name and the actual value, before any ordering check or Polars comparison. Pinned with `test_set_split_dates_rejects_non_date_bounds` parametrised over string, int, float, and `None`.

File: CHANGELOG.md [hunk 7]
```diff
@@ -886,3 +886,7 @@ Note: add all new changelog entries to the bottom of this file.
...
 - Export the canonical outcome targets from `limen.targets` and add direct plus manifest-path regression coverage
+
+## v3.2.0 on 17th of May, 2026
+
+- Add `Manifest.set_split_dates(train_start, train_end, val_start, val_end, test_start, test_end)` and `limen.data.utils.split_by_dates` for absolute-date splits. When `split_dates` is set it takes precedence over `split_config` in both `_run_prepare_setup` (used by `MLManifest.prepare_data` and `RuleBasedManifest.prepare_data`) and `Manifest.compute_test_bars`; otherwise the existing ratio-based `split_sequential` path runs unchanged. Windows are half-open `[start, end)` and the setter validates non-decreasing ordering (`train_start ≤ train_end ≤ val_start ≤ val_end ≤ test_start ≤ test_end`) so non-overlap is mechanical; gaps between adjacent windows are allowed and any rows in those gaps are intentionally excluded from all three splits. Use this when temporal split boundaries must land exactly on the datetime (e.g. honouring a deployment date with no look-ahead leak), in preference to the ratio-based split whose boundary drifts with upstream feature trim.
```
[the remark this answers is not in the document]
Comment (mikkokotila) [reply CHANGELOG.md]: Fixed in a71d480. The CHANGELOG entry no longer claims the ratio split boundary 'drifts with upstream feature trim' — that was wrong because the split actually runs on raw data in `_run_prepare_setup` before feature transforms. The corrected rationale: use `set_split_dates` when temporal boundaries must land on specific datetimes (e.g. honouring a deployment date), in preference to ratio-based splits whose absolute boundary depends on the input row count. The entry also documents the new `with_params_override` clearance contract and the type validation.

File: limen/experiment/manifest_core.py [hunk 8]
```diff
@@ -1307,6 +1379,25 @@ def _run_prepare_setup(
     return split_data, all_datetimes, price_data_for_backtest
 
 
+def _resolve_split(manifest: 'Manifest', raw_data: pl.DataFrame) -> list[pl.DataFrame]:
+
```
[the remark this answers is not in the document]
Comment (mikkokotila) [reply limen/experiment/manifest_core.py:1403]: Fixed in a71d480. The `_resolve_split` docstring now accurately states: 'the split filters `raw_data` by datetime windows so each row's slice assignment is fixed by its datetime value before the per-split feature transforms run'. No claim of feature-trim immunity — only the slice-assignment guarantee, which is what the splitter actually enforces.

File: limen/data/utils/__init__.py [hunk 9]
```diff
@@ -9,6 +10,7 @@
 __all__ = [
     'compute_data_bars',
     'random_slice',
+    'split_by_dates',
```
[the remark this answers is not in the document]
Comment (mikkokotila) [reply limen/data/utils/__init__.py:13]: Fixed in a71d480. Added `split_by_dates` to the public utility table in `docs/Utilities.md` and a full `## `split_by_dates`` section documenting half-open windows, gap exclusion, type validation, the `datetime` column requirement, and the recommendation to prefer `Manifest.set_split_dates` inside the experiment pipeline (since it pairs the splitter with ordering validation and the `with_params_override` clearance contract).

File: limen/experiment/manifest_core.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mikkokotila) [reply limen/experiment/manifest_core.py:446]: Fixed in a71d480. Added a full `### `set_split_dates(...)`` subsection to `docs/Experiment-Manifest.md` (immediately after `set_split_config`), documenting:
- the half-open `[start, end)` window contract
- the ordering rule (`train_start ≤ train_end ≤ val_start ≤ val_end ≤ test_start ≤ test_end`) and that gaps are allowed
- type validation (date/datetime only at the API boundary)
- precedence over `set_split_config` in `prepare_data` and `compute_test_bars`
- the `with_params_override(split_config=...)` clearance contract
- when to prefer it over the ratio split (specific datetimes vs. row-count proportions)

File: limen/data/utils/splits.py [hunk 10]
```diff
@@ -64,6 +64,46 @@ def split_random(data: pl.DataFrame, ratios: Sequence[int], seed: int | None = N
     return [data.sample(fraction=1.0, seed=seed, shuffle=True).slice(start, end - start) for start, end in zip(starts, bounds, strict=True)]
 
 
+def split_by_dates(
+    data: pl.DataFrame,
+    train_start: object, train_end: object,
+    val_start: object, val_end: object,
+    test_start: object, test_end: object,
+) -> list[pl.DataFrame]:
```
[the remark this answers is not in the document]
Comment (mikkokotila) [reply limen/data/utils/splits.py:73]: Fixed in a71d480. `split_by_dates` now annotates every bound as `date` and validates `isinstance(value, date)` at the boundary, raising `TypeError` for strings, ints, floats, and `None` — same contract as `Manifest.set_split_dates`. The docstring also lists the `TypeError` raise. Pinned with `test_split_by_dates_rejects_non_date_bounds` parametrised over multiple bad types.

File: limen/experiment/manifest_core.py (hunk 2 above)
Comment (mikkokotila) [reply limen/experiment/manifest_core.py:233]: Fixed in a71d480. `with_params_override` now clears `split_dates` whenever `split_config` is in the override map — the exact pattern you laid out:

```python
new_manifest.split_config = sc
# Ratio override supersedes a previously-pinned date split.
# Without this, _resolve_split would keep using split_dates and
# the override would silently no-op (e.g. Trainer.train_sensors
# passing (1, 0, 0) to retrain on all available data).
new_manifest.split_dates = None
```

The clearance only fires on the `split_config` branch, so overriding a data-source param (`kline_size`, etc.) leaves `split_dates` in force. Both branches are pinned by tests (`test_with_params_override_split_config_clears_split_dates` and `test_with_params_override_non_split_config_keeps_split_dates`).

**Review by zero-bang — APPROVED** [review]

**Review by bit-mis — APPROVED** [review]

**Review by bit-mis — APPROVED** [review]
